github/messages.go

```diff
@@ -132,17 +137,37 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
 //       // Process payload...
 //     }
 //
-func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	if r.Header.Get("Content-Type") == "application/x-www-form-urlencoded" {
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
+	} else {
+		defer r.Body.Close()
```

Comment:
I think it was a bug with the original implementation. If you look at the docs for `http`: https://golang.org/pkg/net/http/ it's in the example. If you don't close it then the connection remains open: https://stackoverflow.com/questions/33238518/what-could-happen-if-i-dont-close-response-body-in-golang

![image](https://user-images.githubusercontent.com/1034429/29336134-c22519b4-81c1-11e7-8be9-71eed50de0ca.png)